Convert to ES2015 modules

RaumZeit opened the pull request:
This adds `tests/` from [stebogits test branch](https://github.com/stebogit/MarchingSquares.js/tree/tests), applies a new source tree structure, and requires `rollup.js` to create the final bundle(s)...

stebogit committed:
TEST: Added first test

- added input validation to isobands
- added validation tests also to isoContours
- updated readme variable names

RaumZeit committed:
Fix input check for `options`

RaumZeit committed:
TEST: Fix expected output for issues

RaumZeit committed:
Reformat source to comply with ES6 modules and produce bundles with rollupjs

- This changes the sources to ES6 modules format, all located in a sub
  directory src/. To build production bundles with UMD wrappers for
  compatibility with browsers, etc. we use rollup.js.

RaumZeit committed:
Split source files into smaller parts to reduce redundancy

RaumZeit committed:
Add rollup-plugin-uglify

RaumZeit committed:
Add dist/ subdir

RaumZeit committed:
Correct require.js example library path

RaumZeit committed:
Update README.md

RaumZeit committed:
More changes in README.md

RaumZeit committed:
Fix package.json, add ES6 module bundle, update README

RaumZeit committed:
Add QuadTree implementation for fast lookup of cells within given range

stebogit commented on README.md:
```diff
@@ -118,11 +201,11 @@ var data = [
 
-var bandWidth = upperBand - lowerBand;
-var band = MarchingSquaresJS.isoBands(data, lowerBand, bandWidth, options);
+var bandWidth = upperBound - lowerBound;
```
👍 better names

stebogit commented on rollup.config.js:
```diff
@@ -0,0 +1,119 @@
+import { uglify } from 'rollup-plugin-uglify';
+
+var license = "/*!\n* MarchingSquaresJS\n* version 1.2.3\n* https://github.com/RaumZeit/MarchingSquares.js\n*\n* @license GNU Affero General Public License.\n* Copyright (c) 2015-2018 Ronny Lorenz <ronny@tbi.univie.ac.at>\n*/\n\n"
```
> ... `Copyright (c) 2015-2018 Ronny Lorenz` ...

you could insert a definitive
```js
"... Copyright (c) 2015-" + (new Date()).getFullYear()  +" Ronny Lorenz ..."
```

RaumZeit replied:
Thanks for the suggestion! I've done that in my `development` branch now...

stebogit commented on src/options.js:
```diff
@@ -0,0 +1,34 @@
+
+
+/* Compose settings specific to IsoBands algorithm */
+var optIsoBands = Object.assign(
```
I believe `Object.assign()` is an ES6 feature, although it looks like you are not using ES6 as standard.
I would not mix ES4 and ES6 code. If you want to use ES6 (I'd encourage you to do so) you could just use the [Babel plugin for Rollup](https://github.com/rollup/rollup-plugin-babel) to compile the `dist/` code.

RaumZeit replied:
I didn't know about that! For now, I refactored the `options` part and do not require `Object.assign()` anymore!

stebogit commented on src/interpolation.js:
```diff
@@ -0,0 +1,84 @@
+ */
+function linear_ab(a, b, v0, v1) {
+    if (v0 > v1) {
+      var tt = v0;
```
`tt` looks like it's never used.

RaumZeit replied:
this is actually a bug, thanks for identifying it! I've corrected that in my development branch!

stebogit commented on src/interpolation.js:
```diff
@@ -0,0 +1,84 @@
+      if (a < v0) {
+        return (v0 - a) / (b - a);
+      } else {
+        return (v1 - a) / (b - a);
```
here (and in many other places) the `else` is redundant; you could make the code more slim:
```js
if (a < v0) {
    return (v0 - a) / (b - a);
}
return (v1 - a) / (b - a);
```

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+
+
+function extractPolygons(grid, settings) {
+  var areas = [];
```
`areas` looks like it's never used as well

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+
+
+function traceBandPaths(grid, settings) {
+  var areas = [];
```
never used

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+
+
+function traceLinePaths(grid, settings) {
+  var areas = [];
```
`areas` never used

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+              if (typeof cc.edges[enter] !== 'object')
+                break;
+
+              var ee = cc.edges[enter];
```
`ee` already declared

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+              if (typeof cc.edges[enter] !== 'object')
+                break;
+
+              var ee = cc.edges[enter];
```
`ee` was already declared above

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+
+function entry_coordinate(x, y, mode, path) {
+  var k = x;
+  var l = y;
```
why renaming these variables here?

stebogit commented on src/polygons.js:
```diff
@@ -0,0 +1,483 @@
+
+function skip_coordinate(x, y, mode) {
+  var k = x;
+  var l = y;
```
same as  before, unnecessary renaming

stebogit commented on dist/marchingsquares-isolines.js:
```diff
@@ -0,0 +1,908 @@
+
+        break;
+
+      case 15:
```
is there a particular reason why you didn't put these cases in order, like from `0` to `15`?

RaumZeit replied:
yes, the cases are grouped in blocks of similar polygons/iso lines...

stebogit commented on src/main.js:
```diff
@@ -0,0 +1,9 @@
+import isoLines from './isolines.js';
+import isoBands from './isobands.js';
+
+export {
```
To avoid duplicate declaration I believe you should do instead:
```js
export * from './isobands';
export * from './isolines';
export {isoLines as isoContours} from './isolines';
```
while adding `export` to `isolines` and `isobands` functions in their own files ~(keeping the `default export` on the bottom)~:
```js
// isolines.js
export function isolines() { ... }

// isobands.js
export function isobands() { ... }
```

**EDIT**: removing the `default export` from the isolines and isobands files you can actually delete the `main_isolines` and `main_isobands` re-export files

stebogit commented on src/main_isolines.js:
```diff
@@ -0,0 +1,7 @@
+export {
+  isoLines as isoLines,
+  isoLines as isoContours
+}
```
Here should be
```js
export * from './isolines'
export {isoLines as isoContours} from './isolines'
```
once added `export` to `function isolines() {...}` (see above)

**EDIT**: I don't think need this file (see above comments)

stebogit commented on rollup.config.js:
```diff
@@ -0,0 +1,119 @@
+    ]
+  },
+  {
+    input: 'src/main_isobands.js',
```
have you tried using just `input: 'src/isobands.js',`?
I wonder if you actually need `main_isobands.js` and `main_isolines.js` files... 🤔 

**EDIT**: if you export directly the main function and removing the `default export` from `isobands.js` and `isolines.js`, you can use those files as input here:
```js
{
    input: 'src/isobands.js',
    output: {
      extend: true,
...   
```

RaumZeit replied:
I've adopted your suggestions to `export` correctly in the `isobands.js` and `isolines.js` files and got rid of the redundant `main_isobands.js` and `main_isolines.js` entry points.

stebogit commented on src/main_isobands.js:
```diff
@@ -0,0 +1,6 @@
+
+import isoBands from './isobands.js';
+
+export {
```
~`export * from './isobands';` should be enough, no import required~

**EDIT**: you don't need this file (see above comments)

RaumZeit replied:
got rid of it in my development branch! thanks for pointing that out!

stebogit reviewed (commented):
@RaumZeit I just left few comments.

stebogit commented:
@RaumZeit just as a suggestion, you could use a linting tool like [ESLint](https://eslint.org/docs/user-guide/getting-started) and add it to Rollup using its [plugin](https://github.com/TrySound/rollup-plugin-eslint). The linter would help you catch few "errors" like unused or redeclared variables and fix code style inconsistency for you.

If you don't have your own style rules, you could use or extend a popular one like [standardjs](https://standardjs.com/) or just the [`eslint:recommended`](https://eslint.org/docs/rules/) ones.

RaumZeit committed:
Convert QuadTree implementation to something more class-like

RaumZeit committed:
Add options.quadTree property for isoBands

RaumZeit committed:
QuadTree is an object that stores the input data grid and a root node

RaumZeit committed:
Enable use of QuadTree in IsoBands implementation

- This also changes the `isoBands` API to allow for array input in
  `lowerBound` and `bandWidth` parameters. In such cases, te `isoBands`
  function then returns an array of polygon arrays, i.e. one array of
  polygons for each band specified by the input.

RaumZeit committed:
Export quadTree() constructor as MarchingSquaresJS.prepareData() function

- This enables users to pre-process their data, i.e. compute the
  QuadTree, and pass it to subsequent calls of `isoBands` or `isoLines`.
  Therefore, this reduces the overhead of computing the QuadTree for
  each call of the library functions.

RaumZeit committed:
EXAMPLE: Make use of QuadTree optimization in vienna.html

RaumZeit committed:
Generate last year in bundle banner comment automagically

RaumZeit committed:
Refactor input option (user settings) management

RaumZeit committed:
Larger source cleanup and small fixes

- This cleans up some parts of the code, removes redundant main_* files,
  fixes a bug in interpolation.js, adds options.quadTree property to the
  available user settings, and adds some basic input checking for the
  quadTree() constructor.

RaumZeit committed:
DIST: update dist/ directory content

RaumZeit committed:
Add rollup-plugin-eslint to bundle config

RaumZeit committed:
Cleanup src/*.js according to ESLint warnings/errors

RaumZeit commented:
> you could use a linting tool like ESLint ...

Thanks for this suggestion! I've done that now and already started to fix most of the remaining issues.
As soon as the last issues are dealt with, I'll push the changes again.

Also, I've started development of a new feature to substantially speed-up successive isoband/isoline computations. All will be revealed in the next push..

P.S.: When programming in `C` many of the above problems do not appear, since the compiler would always complain about re-declared variables etc. I find the `eslint` plugin very  useful, especially to enforce certain source styling rules! This is something that I usually also use in my `C` projects... Thanks again for hinting in that direction!

RaumZeit committed:
Add QuadTree speed-up feature to isoLines implementation

RaumZeit committed:
Fix some remaining issues in isoBands implementation

- ... and do not export quadTree with isobands.js

RaumZeit committed:
sanitize main.js

RaumZeit committed:
EXAMPLE: Adapt vienna.html to new quadTree feature(s)

RaumZeit committed:
eslint: do not warn on constant conditions

RaumZeit committed:
Fix error messages in isoBands and isoLines

RaumZeit committed:
Adjust test suite to changes in implementation

- Still not entirely sure why issue#6-3_5 was expecting outer frame
  before... It should be correct now, though.

RaumZeit committed:
DIST: update dist/ directory content

RaumZeit commented:
@stebogit Sorry for the large updates in this PR ;)

I've been implementing the performance update on top of the `dev_modules` branch as soon as I created the PR. Thus, it was too difficult to first incorporate your review requests and then rebase my changes to `dev_modules` again. Still, I tried following all the comments you've made earlier!

Anyway, the state of this branch is somewhat more stable now, as all the features I considered urgent are included, and even the test suite has been updated.

I'll still add some more tests and documentation, especially for the new features:
- array input for `threshold` parameter in `isoLines()`
- array input for `lowerBound` and `bandWidth` parameters in `isoBands()`

This already reveals some of the new features:
- the iso line/band limits may be arrays now. In this case, the functions simply return an array, where the top-most level entries are the polygon/line paths for each limit.
- a new options flag `options.quadTree` may be set to speedup consecutive lookup of those regions in the scalar field that actually contribute to iso lines/bands

The latter substantially reduces the computation time, especially when mutliple iso lines/bands are required for the same scalar field, or the iso lines/bands enclose just a small area (the `example/vienna.html` example now only takes 2s instead of 5s on my computer!).

I've also added `ESLint` which helped a lot to find wrong variable declarations, missing semicolons, etc..

RaumZeit committed:
Add .travis.yml and remove package-lock.json

- This triggers Travis CI to automatically build and deploy releases.
  Build status for the master branch is displayed on top of README.md
- We probably do not need the package-lock.json. It only specifies which
  versions of developer tools we used last time to successfully build
  this repo content. Personally, I think it simply spams into the
  git-commits. The solution for now is to remove it and add it to
  .gitignore

RaumZeit committed:
Remove dist/ directory from tracked files

- The contents of dist/ are generated automatically with each build. So
  we remove it and add the entire directory to .gitignore.
  Users who clone the repo must be able to run the build script, e.g.
  `npm run-script build` or `rollup -c`.

RaumZeit committed:
Update README.md

RaumZeit committed:
Use QuadTree optimization automatically if more than one iso line/band is requested

- This transforms the implementation to automatically activate the
  QuadTree optimization for multi line/band input. The user still may
  deactivate the optimization feature using the `options.noQuadTree`
  flag. If the input is a pre-processed scalar field that already
  includes a QuadTree, then the optimization is used even for single
  iso lines/bands.

RaumZeit committed:
EXAMPLE: Adapt vienna.html to changes in `options` object

RaumZeit committed:
Add description for the `options.noQuadTree` flag to README.md

RaumZeit committed:
Add 'npm run-script build' to .travis.yml

RaumZeit committed:
Fix minor spelling mistakes in README.md

RaumZeit committed:
Add description for MarchingSquaresJS.quadTree() function

RaumZeit committed:
EXAMPLE: Simplify vienna.html (again)

RaumZeit committed:
Change order of cell retrieval

- Instead of A -> B -> C -> D we now traverse the quadrants in
  A -> B -> D -> C order.

RaumZeit committed:
Enable the polygon tracers to work on sparse cell list

- This finally allows us to remove all those cells from the MS
  computations that are fully subsumed by the iso line / band limit(s).

RaumZeit committed:
Do not rename functions on export

RaumZeit committed:
Update README.md

RaumZeit committed:
Execute the callback function for each (pair of) limit(s)  and pass the limit(s) as 2nd and 3rd argument(s)

RaumZeit commented:
@stebogit I'll merge this branch into `master` now, if there are any more comments, feel free to open new issues! I'll still give it some more time and testing until the next version will be released...
